### ISSUE TITLE
rebasing

### ISSUE THREAD

**Issue by Keziyah** [Keziyah/fun-git#1]
In which Keziyah learns what a rebase is, and how it can get her out of this situation: 

I'm working on a feature branch, `my-feature-1`, and while it's waiting to be merged, I need to open another branch that depends on it. So while in `my-feature`, I `git checkout -b my-feature-2`. 

I work on my part 2 branch, meanwhile `my-feature` has since been merged, as well as a few other branches that other people have worked on. How do I get `my-feature-2` into master without messing things up?

### What is rebasing

Rebasing and merging have the same end goal. Taking feature branches and integrating them into master. But they do it in different ways. Merges create a new merge commit on master. 
<img width="597" alt="Screen Shot 2020-02-04 at 11 11 54 AM" src="https://user-images.githubusercontent.com/17310089/73730258-29526980-473f-11ea-89cb-0b1cbc04d9da.png">

Rebasing takes you feature commits and moves them on top of the master branch. It rewrites history by erasing your feature branch commits and just putting your feature on top of master. 
![rebase gif](https://user-images.githubusercontent.com/17310089/73730572-a5e54800-473f-11ea-98d5-bc7dadbaf129.gif)

[Source](https://www.youtube.com/watch?v=f1wnYdLEpgI)

### So, like, what do I do

According to [this](https://softwareengineering.stackexchange.com/a/351790) Stack Overflow answer, I should do something like this: 

1. After opening a new branch `my-feature-2` from `my-feature-1`
2. Anytime `my-feature-1` changes during code review, rebase `my-feature-2` on it. 
`... commit something onto my-feature-1 ...
git checkout my-feature-2
git rebase my-feature-1`
This will take my feature 2 commits and put them on top of what i've done from feature 1, making my feature 2 commit history appear to start right after my latest changes to feature 1. 
3. When feature 1 is merged, rebase feature 1 onto it again: 
`git checkout master
git pull origin master
git checkout feature-2
git rebase --onto master feature-1 feature-2`
This makes it so that my feature-2 branch is branched off master. 

### Task
I'm going to try this. 
1) Create a feature-1 branch, make some commits
2) Create a feature-2 branch off feature 1
3) Create other branches and merge them into master before feature-1 is merged
4) Merge feature-1 and make it so that feature-2 is branched off master

## Results
1. I created 4 branches: features 1-4. Feature 2 is branched off Feature 1. 
2. I merged feature-3 and feature-4 into master. 
3. I went into feature-1 and did `git merge master`
4. Since feature-1 has changed, I have to rebase feature-2 on it. So I checked out feature-2 and did `git-rebase feature-1`
Now my feature-2 looks like it began after the latest master merge into feature-1.
#### Before rebasing:  
Feature 2 history: 
<img width="763" alt="Screen Shot 2020-02-04 at 11 54 16 AM" src="https://user-images.githubusercontent.com/17310089/73733689-15116b00-4745-11ea-801e-a8d13bb67c17.png">
Feature 1 history: 
<img width="772" alt="Screen Shot 2020-02-04 at 11 55 29 AM" src="https://user-images.githubusercontent.com/17310089/73733795-4427dc80-4745-11ea-8b1c-66884d988a96.png">
#### After rebasing: 
Feature 2 history
<img width="757" alt="Screen Shot 2020-02-04 at 12 02 14 PM" src="https://user-images.githubusercontent.com/17310089/73734358-3161d780-4746-11ea-9207-65fefdb60a87.png">

Then I merged feature-1 into master. 

Now to clean up the difference between `feature-2` and master. There are more changed files than necessary. The ReadMe, for example, is exactly how it is in master. That's why we need to rebase. 
So while in feature-2, I did `git rebase --onto master feature-1 feature-2`

Changed file list before rebase: 
<img width="1016" alt="Screen Shot 2020-02-04 at 12 07 53 PM" src="https://user-images.githubusercontent.com/17310089/73735064-573bac00-4747-11ea-8cd4-028f5802559c.png">

After rebase: 
Ok, never mind, it didn't work. The changed files in feature-2 are still the same. 

### Second try
1. Created 3 branches. f-1, f-2, and f-3. f-2 depends on f-1. 
2. Merged f-3 into master. 
3. did `git rebase master f-1` while in branch f-1. so that my f-1 commits come after the latest master.
4. Merged f-1 into master
5. git rebase master f-1
6. git rebase --onto feature1 f-1@{1} f-2
[Source](https://coderwall.com/p/xzsr9g/rebasing-dependent-branches-with-git)
I think this worked better. Woohoo!
